File: src/priv/bajdcc/LALR1/grammar/codegen/ICodegen.java
package priv.bajdcc.LALR1.grammar.codegen;

import priv.bajdcc.LALR1.grammar.runtime.RuntimeInst;
import priv.bajdcc.LALR1.grammar.runtime.RuntimeInstBinary;
import priv.bajdcc.LALR1.grammar.runtime.RuntimeInstNon;
import priv.bajdcc.LALR1.grammar.runtime.RuntimeInstUnary;
import priv.bajdcc.LALR1.grammar.tree.Function;

/**
 * 【目标代码生成】接口
 *
 * @author bajdcc
 */
public interface ICodegen {

	void genFuncEntry(String funcName);
	RuntimeInstNon genCode(RuntimeInst inst);
	RuntimeInstUnary genCode(RuntimeInst inst, int op1);
	RuntimeInstUnary genCodeWithFuncWriteBack(RuntimeInst inst, int op1);
	RuntimeInstBinary genCode(RuntimeInst inst, int op1, int op2);
	int genDataRef(Object object);
	int getFuncIndex(Function func);
	int getCodeIndex();
	ICodegenBlock getBlockService();
}

File: src/main/kotlin/com/bajdcc/LALR1/grammar/codegen/ICodegen.kt
package com.bajdcc.LALR1.grammar.codegen

import com.bajdcc.LALR1.grammar.runtime.RuntimeInst
import com.bajdcc.LALR1.grammar.runtime.RuntimeInstBinary
import com.bajdcc.LALR1.grammar.runtime.RuntimeInstNon
import com.bajdcc.LALR1.grammar.runtime.RuntimeInstUnary
import com.bajdcc.LALR1.grammar.tree.Func

/**
 * 【目标代码生成】接口
 *
 * @author bajdcc
 */
interface ICodegen {

    val codeIndex: Int

    val blockService: ICodegenBlock

    fun genFuncEntry(funcName: String)

    fun genCode(inst: RuntimeInst): RuntimeInstNon

    fun genCode(inst: RuntimeInst, op1: Int): RuntimeInstUnary

    fun genCodeWithFuncWriteBack(inst: RuntimeInst, op1: Int): RuntimeInstUnary

    fun genCode(inst: RuntimeInst, op1: Int, op2: Int): RuntimeInstBinary

    fun genDataRef(obj: Any): Int

    fun getFuncIndex(func: Func): Int

    fun genDebugInfo(start: Int, end: Int, info: Any)
}
